### ISSUE TITLE
Creating new calendar items in agenda

### ISSUE THREAD

**Issue by Yenthe666** [Narcolapser/python-o365#35]
Hi @Narcolapser,

First of all, thank you so much for building this library!
I would really like to use this library to create new events in the calendar (with meeting rooms and people attached to it) but I couldn't get it working after hours of trying.

My imports:
```
from O365 import Message
from O365 import Event
from O365 import Calendar
```

The code to create a new calendar item:
```
        e = Event()
        e.setSubject('Coffee!')
        e.setStart(time.gmtime(time.time()+3600)) #start an hour from now.
        e.setEnd(time.gmtime(time.time()+7200)) #end two hours from now.
        new_e = e.create()
```

However, when I run this code, no new calendar item is made in the 365 account. I assume because there is no authentication so the library doesn't know where to create it, so I assigned credentials to it like this:
```
authenticiation = ('my.name@email.cmo','Test123')
e = Event(authentication)
```

Which gives the following error:
```
TypeError: 'tuple' object does not support item assignment
```

So, I then decided to pass data for the JSON too, like this:
```
authenticiation = ('my.name@email.com','Test123')
e = Event(('my.name@email.com', 'Test123'), {'Subject': 'Test', 'Start': '2016-09-09T10:00:00', 'End': '2016-09-09T19:00:00'})
```

Which doesn't throw any errors up but no event is made in the calendar. When I print out e it contains the following values:
```
{'attendees': [],
 'auth': None,
 'calendar': None,
 'json': {'End': '2016-09-30T14:45:59Z',
          'Start': '2016-09-30T13:45:59Z',
          'Subject': 'Coffee!'}}

```

However, after running this line:
```
new_e = e.create()
```

new_e will only contain `False`, which I assume means no event is made.
So, what exactly am I doing wrong?

P.S: I might want to fork this repo and maintain / expand this, so any feedback is very welcome!

Regards,
Yenthe

**Comment by Yenthe666** [comment]
@Narcolapser, trying with another ping.

**Comment by Narcolapser** [comment]
Sorry! you caught me right as I went on vacation. But I think I have the answer to your problem.

Note [here](https://github.com/Narcolapser/python-o365/blob/master/O365/event.py#L48) the order of the arguments in the constructor. Now take a second look at your call:

`e = Event(('my.name@email.com', 'Test123'), {'Subject': 'Test', 'Start': '2016-09-09T10:00:00', 'End': '2016-09-09T19:00:00'})`

You are doing (auth,json) when it calls for (json,auth) the easiest way around it is to just specify the keyword:

`e = Event(auth=('my.name@email.com', 'Test123'), json={'Subject': 'Test', 'Start': '2016-09-09T10:00:00', 'End': '2016-09-09T19:00:00'})`

That should deal with your tuple error. I made those as key word arguments with defaults. This means you can put them in in any order, but only if you specify the names. If you leave off the names, then you have to put them in as order arguments.

Next I foresee your running into [this](https://github.com/Narcolapser/python-o365/blob/master/O365/event.py#L96)

It's a bit clunky, but we have to work with what Microsoft provided. Unfortunately there is not concept of a default calendar for an event. You must specify the calendar. What calendar you say? Again, clunky, but you have to find out what calendars a user has. I reffered to this concept as a "schedule" see my [Vehicle Bookings](https://github.com/Narcolapser/python-o365/blob/master/examples/VehicleBookings/veh.py#L18) example for a little more detail in how that all works.

**Comment by Yenthe666** [comment]
Hi @Narcolapser,

Thanks a lot for your response! Alright I've edited the event in order to first send the auth and then the Json, which works fine.
As you've predicted I am now running into `no valid calendar to upload to.`. I looked at your example but it is using a file as input to read data from so I didn't quite get how to change that to the online API.
I've added my own code with requests to simply get the id of the calendar like this:
```
import requests
calendar_id = None
response = requests.get(calendars_url, auth=('my.name@mycompany.be', 'MyPass'))
for calendar in response.json()['value']:
  // This is the agenda I always need.
  if calendar['Name'] == 'Agenda':
    print(calendar['Id'])
    calendar_id = calendar['Id']
e.calendar = calendar_id
```

So I now have the calendar ID but then setting it to e.calendar doesn't seem to be a solution. It throws me the following error:
```
Traceback (most recent call last):
  File "python-0365.py", line 25, in <module>
    new_e = e.create()
  File "/usr/local/lib/python2.7/dist-packages/O365/event.py", line 93, in create
    calId = self.calendar.calendarId
AttributeError: 'unicode' object has no attribute 'calendarId'
```

So I assume that your library requires an object, not a string, which makes sense. So, how should I pass along the whole object to `e.calendar`? I tried by passing the whole object but it says `AttributeError: 'dict' object has no attribute 'calendarId'` then.

Could you tell me how to either get the calendar record with your library from the 365 or how to pass in the correct object?

Thanks a lot,
Yenthe

**Comment by Narcolapser** [comment]
So does it work for you now?

**Comment by Yenthe666** [comment]
Hi @Narcolapser,

No it doesn't. If I add the whole object the library complains:
```AttributeError: 'unicode' object has no attribute 'calendarId'```
If I add it manually it cannot create the item and I get errors too. See my above answer, it is listed there 💎

**Comment by Yenthe666** [comment]
Hi @Narcolapser,

I believe my previous answer wasn't really clear so let me try to write a new description, which just shows my whole script:
```
import requests
from O365 import Message
from O365 import Event
from O365 import Calendar
from ppretty import ppretty

authenticiation = ('my.email@company.be', 'mypass')
e = Event(('my.email@company.be', 'mypass'), {'Subject': 'Test', 'Start': '2016-10-10T10:00:00', 'End': '2016-10-10T19:00:00', 'attendees': 'my.email@company.be'})
e.attendees = 'my.email@company.be'

# Get a list of all calendars.
calendars_url = 'https://outlook.office365.com/api/v1.0/Me/Calendars'

print('Calendars:')
calendar_fix = None
response = requests.get(calendars_url, auth=('my.email@company.be', 'mypass'))
for calendar in response.json()['value']:
  if calendar['Name'] == 'Agenda':
    print(calendar['Name'])
    print(calendar['Id'])
    calendar_fix = calendar#['Id']
    calendar_fix['CalendarId'] = calendar['Id']
  #print ppretty(calendar)
print(' ')
e.calendar = calendar_fix
new_e = e.create()
print('e: ' + str(e))
print ppretty(e)
```

This will create an object for a new calendar item named 'Test', which is held from 10/10/2016 10:00 up untill  10/10/2016 19:00:00.
However the problem is that when I run this script it will get issues on this line:
```
new_e = e.create()
```

This will throw errors in event.py on line 122:
```
log.error('No response, something is very wrong with create: %s',str(e))
```

When I print out e (which is an event object) like this:
```
print('e: ' + str(e))
```

The result of the object will look like this:
```
e: <O365.event.Event object at 0x7ff921af8c90>
O365.event.Event(
    attendees = 'my.email@company.be', 
    auth = {
            'Start': '2016-10-10T10:00:00', 
            'attendees': 'my.email@company.be', 
            'End': '2016-10-10T19:00:00', 
            'Subject': 'Test'
        }, 
    calendar = {
            u'Name': u'Agenda', 
            u'Color': u'Auto', 
            ..., 
            u'ChangeKey': u'FTOAoV72vU6Vd5cFhXzPQQAAAAAW7g==', 
            u'Id': u'AQMkAGExNTZmMWRmLTZhZDgtNGYyZS1iYTY5LWY2NWRjMWNiZWY5ZABGAAADMc_j2afyWE2J74AdlUfC1wcAFTOAoV72vU6Vd5cFhXzPQQAAAgEGAAAAFTOAoV72vU6Vd5cFhXzPQQAAAhT_AAAA'
        }, 
    isNew = False, 
    json = ('my.email@company.be', 'mypass')
)
```

When I look at the logfile from 365 it looks like this:
```
DEBUG:O365.event:sent to default calendar.
DEBUG:O365.event:creating json for request.
DEBUG:O365.event:sending post request now
ERROR:O365.event:No response, something is very wrong with create: 'dict' object is not callable
```

So there is something wrong with the dictionary or format but I have no idea why/how..
How exactly should I fill that dictionary OR give enough information so it knows where to write the calendar item too?

Again, thank you for the help!
Yenthe

**Comment by Yenthe666** [comment]
@Narcolapser,

I've also added the content from event.py at line 110 in the try catch to know what it contains:(https://github.com/Narcolapser/python-o365/blob/master/O365/event.py#L104-L113)
```
DEBUG:O365.event:sending post request now
ERROR:O365.event:response calId: AQMkAGExNTZmMWRmLTZhZDgtNGYyZS1iYTY5LWY2NWRRWNiZWY5ZABGAAADMc_j2afyWE2J74AdlUfC1wcAFTOAoV72vU6Vd5cFhXzPQQAAAgEGAAAAFTOAoV72vU6Vd5cFhXzPQQAAAhT_AAAA
ERROR:O365.event:response data: ["my.email@company.be", "mypass"]
ERROR:O365.event:response headers: {'Content-type': 'application/json', 'Accept': 'application/json'}
ERROR:O365.event:response auth: {'Start': '2016-10-10T10:00:00', 'attendees': 'my.email@cmopany.be', 'End': '2016-10-10T19:00:00', 'Subject': 'Test'}
```

This is the exact content of the data right before the event is created like this in your library:
```
response = requests.post(self.create_url.format(calId),data,headers=headers,auth=self.auth)
```

**Comment by Narcolapser** [comment]
I think you are making this to complicated. Let me see if I can adapt the vehicle example to your situation:
```python

from O365 import *
import json

auth = ('my.email@company.be', 'mypass')

schedule = Schedule(auth)
schedule.getCalendars()

#this next assumes that you have one calendar and you just want to grab it.
#if you have multiple calendars you'll need to figure out a way to select which
#one you are looking for. This sets 'cal' to a /Calendar Object/ not a dict.
cal = schedule.calendars[0]

#Now we have the calendar, we can create the event. NOTE: Just about everything
#in this library was done with keyword arguments. Very pythonic. And remember:
#If you don't specify what argument goes where, it assumes the order in the
#method declaration. in this case it would have been json, auth, calendar.
#specifically. look at line 8 of the code you just posted see how it breaks this.
e = Event(auth=auth, cal=cal)

#use the accessor methods. they'll save you trouble.
e.setAttendees('my.email@company.be')
e.setSubject('Test')
e.setStart('2016-10-10T10:00:00')
e.setEnd('2016-10-10T19:00:00')

e = e.create()
print('e: ' + str(e))
```

Most of your problems are actually stemming from getting those keyword arguments out of order. Look at your even object print out in your second most recent post. It has the authentication where the json should be and the json where the auth should be. 

NB: If you ever find yourself using requests with this library, you're using it wrong.

**Comment by Yenthe666** [comment]
Hi @Narcolapser,

Thanks a lot for your detailed answer and example!
The main problem is in this line:
```
schedule = Schedule(auth)
```

When I look in your library it calls this URL:
```
cal_url = 'https://outlook.office365.com/api/v1.0/me/calendars'
```

However, when you open this URL and login with valid credentials you will only get a blank page.
![image](https://cloud.githubusercontent.com/assets/6352350/19515748/05e2e3bc-95f9-11e6-91e6-4aa5cbaccf5d.png)
When I run the script I get the following traceback:
```
Traceback (most recent call last):
  File "O365-test-2.py", line 9, in <module>
    schedule.getCalendars()
  File "/usr/local/lib/python2.7/dist-packages/O365/schedule.py", line 36, in getCalendars
    for calendar in response.json()['value']:
  File "/usr/lib/python2.7/dist-packages/requests/models.py", line 741, in json
    return json.loads(self.text, **kwargs)
  File "/usr/lib/python2.7/dist-packages/simplejson/__init__.py", line 488, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/dist-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/usr/lib/python2.7/dist-packages/simplejson/decoder.py", line 389, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
simplejson.scanner.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

Which makes me believe that this API call is outdated and no longer exists?

**Comment by Narcolapser** [comment]
Actually, you just found a bug. I just fixed it with commit [744e944](https://github.com/Narcolapser/python-o365/commit/744e9440e57bedfb534434b02925dcde9a48ed9b). So get the library from from github and that should solve at least that error. 

No offense, but what is your first programming language? I mean, you clearly know how to code, but some python constructs aren't coming natural to you.

I'll gladly help you work on updating and expanding O365. I was sad when I had to move on to other things. I do now have a O365 account with my new company, but it is only a testing account and not my primary, so I've not had incentive to get back into this library. But if I'm helping some one else then I might be more interested.

**Comment by Yenthe666** [comment]
Hi @Narcolapser I actually just found a second one. The import for groups was also missing in your .py file (See PR https://github.com/Narcolapser/python-o365/pull/37).

No offense taken, my first programming languages where C# and Java (with some others such as PHP etc). It is only after I've left college that I started working with Python. Up untill now I've never had the time to dig deeper in things such as python constructors etc. I code for an ERP system but it has its own API, which is all I needed untill now. Constructors are something I really want to learn & understand, it just hasn't come around I guess.
I've never needed them untill now, when you're creating a library or using an existing one like yours it looks like I miss some knowledge there. 😞 

So, now the whole script works fine there is just one issue: when I print out e (which is the result returned from the e.Create function) It will return `None` on the second line:
```
e = e.create()
print pprint(e)
```
The result from pprint:
```
<O365.event.Event object at 0x7effab856d90>
None
```
In the 365 agenda there isn't any new calendar item created but there are no errors being thrown up either. When I dig some further, into the logfile from the library itself I will get this result:
```
DEBUG:O365.event:sent post request.
DEBUG:O365.event:response to event creation: <Response [400]>
```
Any idea what might be wrong or how to progress in this?

**Comment by Narcolapser** [comment]
I was going to guess Java, though C# also makes sense. I've been self taught on python almost from the beginning, though I did Java for school and use Java now professionally. My personal favorite language is still python. Anyway, that helps, I know a little better the point of view you are approaching this with.

Huh. HTTP400: Bad request? Not seen that in a while. can you post y our event object right before you call create? The API may have changed since I first wrote this, that's my guess.

**Comment by Yenthe666** [comment]
Sorry I didn't have time for that anymore yesterday. Here is the whole object:
```
O365.event.Event(
    attendees = [], 
    auth = ('myemail@company.be', 'mypass'), 
    calendar = O365.cal.Calendar(
            auth = ('myemail@company.be', 'mypass'), 
            calendarId = u'AQMkAGExNTZmMWRmLTZhZDgtNGYyZS1iYTY5LWY2NWRjMWNiZWY5ZABGAAADMc_j2afyWE2J74AdlUfC1wcAFTOAoV72vU6Vd5cFhXzPQQAAAgEGAAAAFTOAoV72vU6Vd5cFhXzPQQAAAhT_AAAA', 
            events = [], 
            json = {
                    u'Color': u'Auto', 
                    u'Name': u'Agenda', 
                    u'ChangeKey': u'FTOAoV72vU6Vd5cFhXzPQQAAAAAW7g==', 
                    u'@odata.id': u"https://outlook.office365.com/api/v1.0/Users('myemail@company.be')/Calendars('AQMkAGExNTZmMWRmLTZhZDgtNGYyZS1iYTY5LWY2NWRjMWNiZWY5ZABGAAADMc_j2afyWE2J74AdlUfC1wcAFTOAoV72vU6Vd5cFhXzPQQAAAgEGAAAAFTOAoV72vU6Vd5cFhXzPQQAAAhT_AAAA')", 
                    u'Id': u'AQMkAGExNTZmMWRmLTZhZDgtNGYyZS1iYTY5LWY2NWRjMWNiZWY5ZABGAAADMc_j2afyWE2J74AdlUfC1wcAFTOAoV72vU6Vd5cFhXzPQQAAAgEGAAAAFTOAoV72vU6Vd5cFhXzPQQAAAhT_AAAA'
                }, 
            name = u'Agenda'
        ), 
    json = {
            'Start': '2016-10-10T10:00:00', 
            'End': '2016-10-10T19:00:00', 
            'Attendees': [{'EmailAddress': {...}}], 
            'Subject': 'Test'
        }
)
```

Clearly the problem is in Attendees where there is no data inside, just searching for where I should fill it up.

**Comment by Yenthe666** [comment]
New update, I had to tweak the script in a few ways because the timeformat was not correct & I added two methods for setting the timezone (will create a PR for this later).
I now run into this error:
```
u'error': {
    u'message': u"The property with name '' was found with a value node of type 'PrimitiveValue'; however, a complex value of type 'Microsoft.OutlookServices.Attendee' was expected.", 
    u'code': u'RequestBodyRead'
}
```

The latest script version looks like this:
```
from O365 import *
import json
from ppretty import ppretty

auth = ('myemail@company.be', 'mypassword')

schedule = Schedule(auth)
schedule.getCalendars()

#this next assumes that you have one calendar and you just want to grab it.
#if you have multiple calendars you'll need to figure out a way to select which
#one you are looking for. This sets 'cal' to a /Calendar Object/ not a dict.
cal = schedule.calendars[0]

#Now we have the calendar, we can create the event. NOTE: Just about everything
#in this library was done with keyword arguments. Very pythonic. And remember:
#If you don't specify what argument goes where, it assumes the order in the
#method declaration. in this case it would have been json, auth, calendar.
#specifically. look at line 8 of the code you just posted see how it breaks this.
e = Event(auth=auth, cal=cal)

#use the accessor methods. they'll save you trouble.
e.setAttendee(['myemail@company.be', 'Test'])
e.setSubject('Test')
e.setStart('2016-10-10T18:00:00-08:00')
e.setStartTimeZone('Pacific Standard Time')
e.setEnd('2016-10-10T19:00:00-08:00')
print ppretty(e)
e = e.create()

```

**Comment by Yenthe666** [comment]
@narcolapser one more thing. Any chance the package can be repackaged so all fixes are in the official download?

**Comment by Narcolapser** [comment]
sure, I'll look into it.

**Comment by Yenthe666** [comment]
@Narcolapser any update on this? I'm sorry but I'd like the new version available as installer 😉

**Comment by Narcolapser** [comment]
Got it: https://pypi.python.org/pypi/O365/0.9.1

had to re-figure out how to do that. It's been a while. ^^;

**Comment by Yenthe666** [comment]
Thanks a lot! 👍

**Comment by Yenthe666** [comment]
Hi @Narcolapser,

I found another odd one with this line of code:
```
e.setBody('Test')
```

It generates the following error:
```
 File "/usr/local/lib/python2.7/dist-packages/O365/event.py", line 232, in setBody
    self.json['Body']['Content'] = val
KeyError: 'Body'
```

Coming from this function:
```
	def setBody(self,val):
		'''sets event body content.'''
		self.json['Body']['Content'] = val
```

Any clue?

Regards,
Yenthe

**Comment by Narcolapser** [comment]
Open a new issue. 

and if you can give a little more of your code that would be helpful. setBody works fine for me, so I need to see how your code is different.
